### PR TITLE
Add NSS endpoints for ECH interop

### PR DESCRIPTION
**Pull request by kjacobs-moz**
This currently works for `ech-accept` between NSS and cloudflare-go.

### CODE REVIEW

**Comment by cjpatton** [comment]
Is this ready to review, @kjacobs-moz?

**Comment by kjacobs-moz** [comment]
> Is this ready to review, @kjacobs-moz?

Yes, the NSS change has landed.

File: impl-endpoints/nss/ech_key_converter.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+ECH_VERSION = 0xFE09
+DHKEM_X25519_SHA256 = 0x0020
+pkcs8_start = b"\x30\x67\x02\x01\x00\x30\x14\x06\x07\x2A\x86\x48\xCE\x3D\x02\x01\x06\x09\x2B\x06\x01\x04\x01\xDA\x47\x0F\x01\x04\x4C\x30\x4A\x02\x01\x01\x04\x20"
+pkcs8_oid = b"\xa1\x23\x03\x21\x00"
+
+
+def convert_ech_key(in_file, out_file):
+    with open(in_file, "rb") as f:
+        ech_keys = base64.b64decode(f.read(), None, True)
```
Comment (cjpatton) [inline impl-endpoints/nss/ech_key_converter.py:13]: Suggest s/ech_keys/ech_keypair/. ech_keys sounds like it's a sequence of ECHKey structures.

File: impl-endpoints/nss/run_endpoint.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+mkdir nss_testdata
+
+if [ "$TESTCASE" = "ech-accept" ]; then
+    # Create a PKCS8 file for the ECH keypair
+    python3 /ech_key_converter.py testdata/ech_key nss_testdata/ech_key_converted
+
+    # Create pfx files for pk12util
+	openssl pkcs12 -export -out nss_testdata/root.pfx -name root.com -inkey testdata/root.key -in testdata/root.crt -passout pass:"$P12_PASS"
+	openssl pkcs12 -export -out nss_testdata/example.pfx -name example.com -inkey testdata/example.key -in testdata/example.crt -passout pass:"$P12_PASS"
+	openssl pkcs12 -export -out nss_testdata/client-facing.pfx -name client-facing.com -inkey testdata/client-facing.key -in testdata/client-facing.crt -passout pass:"$P12_PASS"
```
Comment (cjpatton) [inline impl-endpoints/nss/run_endpoint.sh:20]: nit: Align these lines with the indentation of the block.

File: impl-endpoints/nss/ech_key_converter.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import sys
+import struct
+import base64
+
+ECH_VERSION = 0xFE09
+DHKEM_X25519_SHA256 = 0x0020
+pkcs8_start = b"\x30\x67\x02\x01\x00\x30\x14\x06\x07\x2A\x86\x48\xCE\x3D\x02\x01\x06\x09\x2B\x06\x01\x04\x01\xDA\x47\x0F\x01\x04\x4C\x30\x4A\x02\x01\x01\x04\x20"
+pkcs8_oid = b"\xa1\x23\x03\x21\x00"
```
Comment (cjpatton) [inline impl-endpoints/nss/ech_key_converter.py:8]: How does NSS Interpret these values? Are they specific to X25519? If so, please add a comment explaining that they are.

File: cmd/util/make.go [hunk 4]
```diff
@@ -106,7 +106,7 @@ func makeIntermediateCertificate(config *Config, inCertPath string, inKeyPath st
 		SerialNumber: serialNumber,
 		Subject: pkix.Name{
 			Organization:       []string{"tls-interop-runner development certificate"},
-			OrganizationalUnit: []string{"tls-interop-runner"},
+			OrganizationalUnit: []string{outPath},
```
Comment (cjpatton) [inline cmd/util/make.go:109]: <3

**Review by cjpatton — CHANGES_REQUESTED** [review]
This looks great! Just a few minor nits and one ask (see inline comments).

Also, a question: The ECH draft does not specify how to format secret keys. Would you endorse PKCS#8 as a format for our purposes here? In other words, should we modify `util.go` to output an ECHConfigs and, separately, an the corresponding PKCS#8-formatted key pair? (See #17.)

**Comment by cjpatton** [comment]
> For the other direction, I'll need to land a change to [selfserv.c](https://searchfox.org/mozilla-central/source/security/nss/cmd/selfserv/selfserv.c#1967) to consume the PKCS8-format HPKE keypair. With that change applied, connection succeeds from cloudflare-go to NSS.

FWIW, cloudflare-go client -> nss server works correctly with the existing conversion script.

**Review by chris-wood — APPROVED** [review]
LGTM pending @cjpatton's suggestions for the key conversion script.

**Comment by cjpatton** [comment]
> LGTM pending @cjpatton's suggestions for the key conversion script.

I'm happy to merge without the conversion and get back to it later.

**Comment by chris-wood** [comment]
Yeah, same. I was referring to the documentation suggestions. But even those I'd be fine without.

**Comment by kjacobs-moz** [comment]
> > LGTM pending @cjpatton's suggestions for the key conversion script.
> 
> I'm happy to merge without the conversion and get back to it later.

Thanks for the reviews.

Unfortunately, I'm not going to have time in the immediate future to rewrite it in Go, but I would definitely support PKCS8-formatting the ECH/HPKE keypair rather than storing the raw private key, which some libraries may not be able to import easily. NSS is one such library, and doing it outside of NSS/selfserv avoids having to write a second ECHConfigs parser at the application level. With that change, we could remove the script entirely, but it might be worth waiting to see if other libraries have an opinion.

Another option is to output two formats from util.go.

**Review by cjpatton — APPROVED** [review]
LGTM. Pleas squash the last commit before merging.

**Comment by cjpatton** [comment]
> LGTM. Pleas squash the last commit before merging.

Oops, scratch that. Commits look good. I'm merging now.

**Pull request merged**